File: server/learning.js
import { createHash, randomUUID } from 'node:crypto';
import { resolvePolicy } from './policy.js';
import { monthsAgo } from './history.js';
import { modelPayload } from './integrations.js';

const defaults = { enabled: false, weekly: false, months: 3, maxSamples: 50, tokenBudget: 16000 };
const fail = (message, status = 409) => { throw Object.assign(new Error(message), { status }); };
const hash = value => createHash('sha256').update(JSON.stringify(value)).digest('hex');
// ponytail: deterministic quote/signature heuristics; review samples because mail formats vary.
export function ownText(body = '') {
  return body.replace(/\r\n?/g, '\n').split('\n').filter(line => !/^\s*>/.test(line)).join('\n')
    .split(/\n(?:On .{3,200}wrote:|在.{3,200}(?:寫道|写道)[：:]|[-_ ]{2,}(?:Original Message|Forwarded message|原始郵件|轉寄郵件)|From:\s|寄件者[：:]|--\s*$|Sent from my |Get Outlook for |CONFIDENTIAL(?:ITY)?\b|DISCLAIMER\b|免責聲明|保密聲明)/im)[0].trim();
}
export function createLearning({ store, connection, runModel, now = Date.now }) {
  const read = account => ({ settings: defaults, ...store.getSettings().styleLearning?.[account] });
  const write = (account, value) => store.setSettings({ styleLearning: { ...store.getSettings().styleLearning, [account]: value } });
  const config = account => ({ ...defaults, ...read(account).settings });
  function permitted(account) {
    const policy = resolvePolicy(store.getSettings().policy);
    return !!connection(account) && config(account).enabled && policy.enabled && policy.behaviors.memory && policy.folders.sent && policy.content.body;
  }
  function stamp(account) {
    const settings = store.getSettings();
    return hash([connection(account)?.connectionId || connection(account)?.email, settings.ai, resolvePolicy(settings.policy), settings.preferences, config(account)]);
  }
  function source(account, id) {
    const message = store.getMessage(account, id);
    return message && message.folder === 'sent' && message.fromEmail?.toLowerCase() === account.toLowerCase() && !message.automated ? ownText(message.body) : null;
  }
  function valid(account, preview) {
    return permitted(account) && preview && preview.stamp === stamp(account) && preview.sources.every(item => hash(source(account, item.id)) === item.hash);
  }
  function updateSettings(account, input) {
    if (!connection(account)) fail('Choose a connected mailbox.');
    if (!input || typeof input !== 'object' || Array.isArray(input) || Object.keys(input).some(key => !Object.hasOwn(defaults, key))) fail('Invalid style settings.', 400);
    const next = { ...config(account), ...input };
    if (typeof next.enabled !== 'boolean' || typeof next.weekly !== 'boolean' || ![1, 3, 6, 12].includes(next.months) || !Number.isInteger(next.maxSamples) || next.maxSamples < 1 || next.maxSamples > 50 || !Number.isInteger(next.tokenBudget) || next.tokenBudget < 4000 || next.tokenBudget > 64000) fail('Choose 1–50 samples and a 4,000–64,000 token budget.', 400);
    if (next.weekly && !next.enabled) fail('Enable style learning before weekly updates.', 400);
    const current = read(account);
    write(account, { ...current, settings: next, preview: null, lastWeeklyAt: now(), weeklySince: next.weekly && !config(account).weekly ? new Date(now()).toISOString() : current.weeklySince });
  }
  const modelOptions = () => ({ preferences: store.getSettings().preferences || {}, includeUsage: true });
  const modelSettings = () => ({ ...store.getSettings().ai, maxTokens: Math.min(store.getSettings().ai?.maxTokens || 1200, 1200) });
  // UTF-8 bytes plus framing is deliberately conservative, not a model-specific tokenizer.
  const estimate = messages => Buffer.byteLength(JSON.stringify(modelPayload(modelSettings(), 'style', messages, '', modelOptions()))) + 256 + modelSettings().maxTokens;
  function prepare(account, incremental = false) {
    if (!permitted(account)) fail('Enable style learning, Email Brain, Sent folder and email body access in AI permissions.', 403);
    const current = read(account), options = config(account), settings = store.getSettings();
    if (!settings.ai?.baseUrl || !settings.ai?.model) fail('Configure an AI model first.');
    if (current.preview?.status === 'running') fail('Style analysis is already running.');
    const end = new Date(now()).toISOString(), start = monthsAgo(options.months, now());
    const since = incremental ? [start, current.analyzedThrough, current.weeklySince].filter(Boolean).sort().at(-1) : start;
    const unique = new Map();
    for (const message of store.listMessages(account)) {
      if (message.date < since || message.date >= end) continue;
      const body = source(account, message.id);
      if (!body || body.length < 40) continue;
      const digest = hash(body.toLowerCase().replace(/\s+/g, ' '));
      if (!unique.has(digest)) unique.set(digest, { message, body });
    }
    // Round-robin date/recipient buckets avoid choosing only one conversation.
    const buckets = new Map();
    for (const item of unique.values()) {
      const key = `${item.message.date.slice(0, 7)}:${item.message.to?.toLowerCase() || ''}`;
      if (!buckets.has(key)) buckets.set(key, []);
      buckets.get(key).push(item);
    }
    const ordered = [];
    while ([...buckets.values()].some(items => items.length)) for (const items of buckets.values()) if (items.length) ordered.push(items.shift());
    const sources = [], messages = [], cap = Math.min(options.maxSamples, resolvePolicy(settings.policy).maxMessages, 50);
    for (const item of ordered) {
      if (sources.length >= cap) break;
      const candidate = { body: item.body.slice(0, 6000) };
      if (estimate([...messages, candidate]) > options.tokenBudget) continue;
      messages.push(candidate); sources.push({ id: item.message.id, hash: hash(item.body), body: candidate.body });
    }
    if (!sources.length) fail('No useful Sent samples fit your dates, permissions and budget. Import Sent mail or increase the budget.');
    const preview = { id: randomUUID(), status: 'prepared', createdAt: end, through: end, incremental, stamp: stamp(account), sources, eligible: unique.size, sampleCount: sources.length, effectiveCap: cap, estimatedTokens: estimate(messages), tokenBudget: options.tokenBudget };
    write(account, { ...current, preview });
    return preview;
  }
  async function generate(account, id) {
    const preview = read(account).preview;
    if (!valid(account, preview) || preview.id !== id || preview.status !== 'prepared') fail('Preview expired or was already used. Prepare fresh samples.');
    write(account, { ...read(account), preview: { ...preview, status: 'running' } });
    try {
      const result = await runModel(modelSettings(), 'style', preview.sources.map(item => ({ body: item.body })), '', modelOptions());
      if (read(account).preview?.id !== id || !valid(account, preview)) fail('Permissions, account, model or source mail changed. Result discarded.');
      const voice = (typeof result === 'string' ? result : result?.text)?.trim();
      if (!voice || voice.length > 2000) fail('The model returned an invalid style guide. Try again with fresh samples.');
      write(account, { ...read(account), analyzedThrough: preview.through, preview: { ...preview, status: 'ready', voice, usage: result.usage || {} } });
    } catch (error) {
      if (read(account).preview?.id === id) write(account, { ...read(account), preview: { ...preview, status: 'failed', error: 'Analysis failed or its context changed. Tokens may have been used. Prepare a new preview to retry explicitly.' } });
      throw Object.assign(new Error('Style analysis failed or its context changed. Check settings and prepare fresh samples; tokens may have been used.'), { status: 502 });
    }
  }
  function apply(account, input) {
    const value = read(account), preview = value.preview;
    if (!valid(account, preview) || preview.id !== input?.previewId || preview.status !== 'ready') fail('Prepare and analyze fresh samples before saving.');
    if (typeof input.voice !== 'string' || !input.voice.trim() || input.voice.length > 2000) fail('Style guide must contain 1–2000 characters.', 400);
    write(account, { ...value, profile: { voice: input.voice.trim(), updatedAt: new Date(now()).toISOString(), sources: preview.sources.map(({ id, hash }) => ({ id, hash })) }, preview: null });
  }
  function voice(account) {
    const profile = read(account).profile;
    return permitted(account) && profile && profile.sources.every(item => hash(source(account, item.id)) === item.hash) ? profile.voice : '';
  }
  function state(account) {
    const value = read(account), preview = value.preview;
    const visible = preview && valid(account, preview);
    return { settings: config(account), permitted: permitted(account), profile: value.profile ? { voice: value.profile.voice, updatedAt: value.profile.updatedAt, active: !!voice(account) } : null,
      preview: visible ? { ...Object.fromEntries(['id', 'status', 'createdAt', 'incremental', 'eligible', 'sampleCount', 'effectiveCap', 'estimatedTokens', 'tokenBudget', 'voice', 'usage', 'error'].map(key => [key, preview[key]])), samples: preview.sources.map(item => ({ body: item.body })) } : null, lastWeeklyAt: value.lastWeeklyAt || null };
  }
  for (const [account, value] of Object.entries(store.getSettings().styleLearning || {})) {
    if (value.preview?.status === 'running') write(account, { ...value, preview: { ...value.preview, status: 'interrupted', error: 'Interrupted by shutdown. Tokens may have been used. Prepare new samples to retry.' } });
  }
  async function tick() {
    for (const account of Object.keys(store.getSettings().styleLearning || {})) {
      const value = read(account);
      if (!permitted(account) || !config(account).weekly || now() < (value.lastWeeklyAt || now()) + 7 * 86400000 || (valid(account, value.preview) && ['prepared', 'running', 'ready'].includes(value.preview?.status))) continue;
      write(account, { ...value, lastWeeklyAt: now() }); // Claim before model use; never replay after a crash.
      try { const preview = prepare(account, true); await generate(account, preview.id); } catch { /* Persisted failure or no new samples; retry next week. */ }
      break;
    }
  }
  return { updateSettings, prepare, generate, apply, voice, state, tick, clear: account => write(account, { settings: { ...config(account), enabled: false, weekly: false } }) };
}

File: rust/src/learning.rs
use crate::{
    ai::{hash, model_payload, run_model, truncate},
    error::{Error, Result},
    policy,
    service::{App, Context, connections},
    store::{Store, merge, now, string},
};
use axum::{
    Json,
    response::{IntoResponse, Response},
};
use chrono::{DateTime, Months, SecondsFormat, Utc};
use regex::Regex;
use serde_json::{Value, json};
use std::{
    collections::{HashMap, HashSet, VecDeque},
    sync::LazyLock,
};

fn defaults() -> Value {
    json!({"enabled":false,"weekly":false,"months":3,"maxSamples":50,"tokenBudget":16000})
}
fn config(settings: &Value, owner: &str) -> Value {
    merge(defaults(), &settings["styleLearning"][owner]["settings"])
}
fn read(settings: &Value, owner: &str) -> Value {
    merge(
        json!({"settings":defaults()}),
        &settings["styleLearning"][owner],
    )
}
fn write(db: &Store, owner: &str, value: Value) -> Result<()> {
    let mut learning = db.settings()?["styleLearning"]
        .as_object()
        .cloned()
        .unwrap_or_default();
    learning.insert(owner.into(), value);
    db.set_settings(&json!({"styleLearning":learning}))?;
    Ok(())
}
fn permitted(settings: &Value, owner: &str) -> bool {
    let policy = policy::resolve(&settings["policy"]);
    connections(settings).get(owner).is_some()
        && config(settings, owner)["enabled"] == true
        && policy["enabled"] == true
        && policy["behaviors"]["memory"] == true
        && policy["folders"]["sent"] == true
        && policy["content"]["body"] == true
}
fn stamp(settings: &Value, owner: &str) -> String {
    let connections = connections(settings);
    let connection = &connections[owner];
    hash(&json!([
        connection
            .get("connectionId")
            .unwrap_or(&connection["email"]),
        settings["ai"],
        policy::resolve(&settings["policy"]),
        settings["preferences"],
        config(settings, owner)
    ]))
}
// ponytail: deterministic quote/signature heuristics; samples remain reviewable because mail formats vary.
pub fn own_text(body: &str) -> String {
    static QUOTE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^\s*>").unwrap());
    static CUT: LazyLock<Regex> = LazyLock::new(|| {
        Regex::new(r"(?im)\n(?:On .{3,200}wrote:|在.{3,200}(?:寫道|写道)[：:]|[-_ ]{2,}(?:Original Message|Forwarded message|原始郵件|轉寄郵件)|From:\s|寄件者[：:]|--\s*$|Sent from my |Get Outlook for |CONFIDENTIAL(?:ITY)?\b|DISCLAIMER\b|免責聲明|保密聲明)").unwrap()
    });
    let text = body
        .replace("\r\n", "\n")
        .replace('\r', "\n")
        .split('\n')
        .filter(|line| !QUOTE.is_match(line))
        .collect::<Vec<_>>()
        .join("\n");
    CUT.split(&text).next().unwrap_or("").trim().to_owned()
}
fn source(db: &Store, owner: &str, id: &str) -> Result<Value> {
    Ok(match db.get(owner, id)? {
        Some(m)
            if m["folder"] == "sent"
                && string(&m, "fromEmail").eq_ignore_ascii_case(owner)
                && m["automated"] != true =>
        {
            json!(own_text(string(&m, "body")))
        }
        _ => Value::Null,
    })
}
fn sources_valid(db: &Store, owner: &str, sources: &Value) -> Result<bool> {
    let Some(items) = sources.as_array() else {
        return Ok(false);
    };
    for item in items {
        if item["hash"] != hash(&source(db, owner, string(item, "id"))?) {
            return Ok(false);
        }
    }
    Ok(true)
}
fn valid(db: &Store, settings: &Value, owner: &str, preview: &Value) -> Result<bool> {
    Ok(permitted(settings, owner)
        && preview.is_object()
        && preview["stamp"] == stamp(settings, owner)
        && preview
            .get("generation")
            .is_none_or(|v| *v == settings["aiGeneration"])
        && sources_valid(db, owner, &preview["sources"])?)
}
pub fn voice(db: &Store, settings: &Value, owner: &str) -> Result<String> {
    let profile = &settings["styleLearning"][owner]["profile"];
    Ok(
        if permitted(settings, owner)
            && profile.is_object()
            && sources_valid(db, owner, &profile["sources"])?
        {
            string(profile, "voice").into()
        } else {
            String::new()
        },
    )
}
/// Conservative projection when source data is unavailable. Use state_with_store for the UI.
pub fn state(settings: &Value, owner: &str) -> Value {
    let value = read(settings, owner);
    let profile = &value["profile"];
    json!({"settings":config(settings,owner),"permitted":permitted(settings,owner),"profile":if profile.is_object(){json!({"voice":profile["voice"],"updatedAt":profile["updatedAt"],"active":false})}else{Value::Null},"preview":null,"lastWeeklyAt":value.get("lastWeeklyAt").cloned().unwrap_or(Value::Null)})
}
pub fn state_with_store(db: &Store, settings: &Value, owner: &str) -> Result<Value> {
    let mut state = state(settings, owner);
    let value = read(settings, owner);
    let preview = &value["preview"];
    if state["profile"].is_object() {
        state["profile"]["active"] = (!voice(db, settings, owner)?.is_empty()).into();
    }
    if valid(db, settings, owner, preview)? {
        let mut visible = json!({});
        for key in [
            "id",
            "status",
            "createdAt",
            "incremental",
            "eligible",
            "sampleCount",
            "effectiveCap",
            "estimatedTokens",
            "tokenBudget",
            "voice",
            "usage",
            "error",
        ] {
            if let Some(value) = preview.get(key) {
                visible[key] = value.clone();
            }
        }
        visible["samples"] = json!(
            preview["sources"]
                .as_array()
                .into_iter()
                .flatten()
                .map(|i| json!({"body":i["body"]}))
                .collect::<Vec<_>>()
        );
        state["preview"] = visible;
    }
    Ok(state)
}
pub fn initialize(db: &Store) -> Result<()> {
    let settings = db.settings()?;
    for (owner, value) in settings["styleLearning"].as_object().into_iter().flatten() {
        if value["preview"]["status"] == "running" {
            let mut value = value.clone();
            value["preview"]["status"] = "interrupted".into();
            value["preview"]["error"] =
                "Interrupted by shutdown. Tokens may have been used. Prepare new samples to retry."
                    .into();
            write(db, owner, value)?;
        }
    }
    Ok(())
}
pub fn update_settings(db: &Store, owner: &str, input: &Value) -> Result<()> {
    update_settings_at(db, owner, input, Utc::now())
}
pub fn update_settings_at(
    db: &Store,
    owner: &str,
    input: &Value,
    time: DateTime<Utc>,
) -> Result<()> {
    let settings = db.settings()?;
    if connections(&settings).get(owner).is_none() {
        return Err(Error::conflict("Choose a connected mailbox."));
    }
    if !input
        .as_object()
        .is_some_and(|map| map.keys().all(|key| defaults().get(key).is_some()))
    {
        return Err(Error::invalid("Invalid style settings."));
    }
    let previous = config(&settings, owner);
    let next = merge(previous.clone(), input);
    if !next["enabled"].is_boolean()
        || !next["weekly"].is_boolean()
        || !next["months"]
            .as_u64()
            .is_some_and(|n| [1, 3, 6, 12].contains(&n))
        || !next["maxSamples"]
            .as_u64()
            .is_some_and(|n| (1..=50).contains(&n))
        || !next["tokenBudget"]
            .as_u64()
            .is_some_and(|n| (4000..=64000).contains(&n))
    {
        return Err(Error::invalid(
            "Choose 1–50 samples and a 4,000–64,000 token budget.",
        ));
    }
    if next["weekly"] == true && next["enabled"] != true {
        return Err(Error::invalid(
            "Enable style learning before weekly updates.",
        ));
    }
    let mut current = read(&settings, owner);
    if next["weekly"] == true && previous["weekly"] != true {
        current["weeklySince"] = time.to_rfc3339_opts(SecondsFormat::Millis, true).into();
    }
    current["settings"] = next;
    current["preview"] = Value::Null;
    current["lastWeeklyAt"] = time.timestamp_millis().into();
    write(db, owner, current)
}
fn model_settings(settings: &Value) -> Value {
    merge(
        settings["ai"].clone(),
        &json!({"maxTokens":settings["ai"]["maxTokens"].as_u64().unwrap_or(1200).min(1200)}),
    )
}
fn model_options(settings: &Value) -> Value {
    json!({"preferences":settings.get("preferences").cloned().unwrap_or(json!({})),"includeUsage":true})
}
fn estimate(settings: &Value, messages: &[Value]) -> Result<usize> {
    let ai = model_settings(settings);
    Ok(serde_json::to_vec(&model_payload(
        &ai,
        "style",
        messages,
        "",
        &model_options(settings),
    )?)?
    .len()
        + 256
        + ai["maxTokens"].as_u64().unwrap_or(1200) as usize)
}
pub fn prepare(db: &Store, owner: &str, incremental: bool) -> Result<Value> {
    prepare_at(db, owner, incremental, Utc::now())
}
pub fn prepare_at(
    db: &Store,
    owner: &str,
    incremental: bool,
    time: DateTime<Utc>,
) -> Result<Value> {
    let settings = db.settings()?;
    if !permitted(&settings, owner) {
        return Err(Error::new(
            403,
            "Enable style learning, Email Brain, Sent folder and email body access in AI permissions.",
        ));
    }
    let current = read(&settings, owner);
    let options = config(&settings, owner);
    if string(&settings["ai"], "baseUrl").is_empty() || string(&settings["ai"], "model").is_empty()
    {
        return Err(Error::conflict("Configure an AI model first."));
    }
    if current["preview"]["status"] == "running" {
        return Err(Error::conflict("Style analysis is already running."));
    }
    let end = time.to_rfc3339_opts(SecondsFormat::Millis, true);
    let start = time
        .checked_sub_months(Months::new(options["months"].as_u64().unwrap_or(3) as u32))
        .ok_or_else(|| Error::invalid("Invalid learning date."))?
        .to_rfc3339_opts(SecondsFormat::Millis, true);
    let since = if incremental {
        [
            start.as_str(),
            string(&current, "analyzedThrough"),
            string(&current, "weeklySince"),
        ]
        .into_iter()
        .max()
        .unwrap_or(&start)
    } else {
        &start
    };
    let mut seen = HashSet::new();
    let mut buckets: Vec<VecDeque<(Value, String)>> = Vec::new();
    let mut positions = HashMap::<String, usize>::new();
    for m in db.list(owner)? {
        if string(&m, "date") < since || string(&m, "date") >= end.as_str() {
            continue;
        }
        let body = source(db, owner, string(&m, "id"))?;
        let Some(body) = body
            .as_str()
            .filter(|body| body.encode_utf16().count() >= 40)
        else {
            continue;
        };
        let normalized = body
            .to_lowercase()
            .split_whitespace()
            .collect::<Vec<_>>()
            .join(" ");
        if !seen.insert(hash(&json!(normalized))) {
            continue;
        }
        let key = format!(
            "{}:{}",
            truncate(string(&m, "date"), 7),
            string(&m, "to").to_lowercase()
        );
        let index = *positions.entry(key).or_insert_with(|| {
            buckets.push(VecDeque::new());
            buckets.len() - 1
        });
        buckets[index].push_back((m.clone(), body.into()));
    }
    let cap = options["maxSamples"]
        .as_u64()
        .unwrap_or(50)
        .min(
            policy::resolve(&settings["policy"])["maxMessages"]
                .as_u64()
                .unwrap_or(8),
        )
        .min(50) as usize;
    let mut sources = Vec::new();
    let mut messages = Vec::new();
    'samples: loop {
        let mut any = false;
        for bucket in &mut buckets {
            let Some((m, body)) = bucket.pop_front() else {
                continue;
            };
            any = true;
            if sources.len() >= cap {
                break 'samples;
            }
            let candidate = json!({"body":truncate(&body,6000)});
            messages.push(candidate.clone());
            if estimate(&settings, &messages)?
                > options["tokenBudget"].as_u64().unwrap_or(16000) as usize
            {
                messages.pop();
                continue;
            }
            sources.push(json!({"id":m["id"],"hash":hash(&json!(body)),"body":candidate["body"]}));
        }
        if !any {
            break;
        }
    }
    if sources.is_empty() {
        return Err(Error::conflict(
            "No useful Sent samples fit your dates, permissions and budget. Import Sent mail or increase the budget.",
        ));
    }
    let preview = json!({"id":uuid::Uuid::new_v4().to_string(),"status":"prepared","createdAt":end,"through":end,"incremental":incremental,"stamp":stamp(&settings,owner),"generation":settings["aiGeneration"],"sources":sources,"eligible":seen.len(),"sampleCount":sources.len(),"effectiveCap":cap,"estimatedTokens":estimate(&settings,&messages)?,"tokenBudget":options["tokenBudget"]});
    write(db, owner, merge(current, &json!({"preview":preview})))?;
    Ok(preview)
}
pub async fn generate(app: &App, owner: &str, id: &str) -> Result<()> {
    let owner = owner.to_owned();
    let id = id.to_owned();
    let account = owner.clone();
    let expected = id.clone();
    let (preview, settings) = app
        .db(move |db| {
            let settings = db.settings()?;
            let current = read(&settings, &account);
            let preview = current["preview"].clone();
            if !valid(db, &settings, &account, &preview)?
                || preview["id"] != expected
                || preview["status"] != "prepared"
            {
                return Err(Error::conflict(
                    "Preview expired or was already used. Prepare fresh samples.",
                ));
            }
            write(
                db,
                &account,
                merge(
                    current,
                    &json!({"preview":merge(preview.clone(),&json!({"status":"running"}))}),
                ),
            )?;
            Ok((preview, settings))
        })
        .await?;
    let messages: Vec<_> = preview["sources"]
        .as_array()
        .into_iter()
        .flatten()
        .map(|i| json!({"body":i["body"]}))
        .collect();
    let result = run_model(
        &app.0.client,
        &model_settings(&settings),
        "style",
        &messages,
        "",
        &model_options(&settings),
    )
    .await;
    app.db(move|db|{
        let settings=db.settings()?;let current=read(&settings,&owner);
        let success=result.ok().filter(|result|{let voice=string(result,"text").trim();!voice.is_empty()&&voice.encode_utf16().count()<=2000});
        if let Some(result)=success.filter(|_|current["preview"]["id"]==id&&current["preview"]["status"]=="running") && valid(db,&settings,&owner,&preview)?{write(db,&owner,merge(current,&json!({"analyzedThrough":preview["through"],"preview":merge(preview.clone(),&json!({"status":"ready","voice":string(&result,"text").trim(),"usage":result["usage"]}))})))?;return Ok(());}
        if current["preview"]["id"]==id{write(db,&owner,merge(current,&json!({"preview":merge(preview,&json!({"status":"failed","error":"Analysis failed or its context changed. Tokens may have been used. Prepare a new preview to retry explicitly."}))})))?;}
        Err(Error::new(502,"Style analysis failed or its context changed. Check settings and prepare fresh samples; tokens may have been used."))
    }).await
}
pub fn apply(db: &Store, owner: &str, input: &Value) -> Result<()> {
    let settings = db.settings()?;
    let value = read(&settings, owner);
    let preview = &value["preview"];
    if !valid(db, &settings, owner, preview)?
        || preview["id"] != input["previewId"]
        || preview["status"] != "ready"
    {
        return Err(Error::conflict(
            "Prepare and analyze fresh samples before saving.",
        ));
    }
    let voice = crate::validation::text(&input["voice"], "Style guide", 2000, false)?.trim();
    let profile = json!({"voice":voice,"updatedAt":now(),"sources":preview["sources"].as_array().into_iter().flatten().map(|i|json!({"id":i["id"],"hash":i["hash"]})).collect::<Vec<_>>()});
    write(
        db,
        owner,
        merge(value, &json!({"profile":profile,"preview":null})),
    )
}
pub fn clear(db: &Store, owner: &str) -> Result<()> {
    let settings = db.settings()?;
    if !crate::service::valid_account(&settings, owner) {
        return Err(Error::conflict("Choose a connected mailbox."));
    }
    write(
        db,
        owner,
        json!({"settings":merge(config(&settings,owner),&json!({"enabled":false,"weekly":false}))}),
    )
}
pub async fn scheduled_tick(app: &App) -> Result<()> {
    let work = app
        .db(|db| {
            let settings = db.settings()?;
            let time = Utc::now();
            for (owner, _) in settings["styleLearning"].as_object().into_iter().flatten() {
                let value = read(&settings, owner);
                if !permitted(&settings, owner)
                    || config(&settings, owner)["weekly"] != true
                    || time.timestamp_millis()
                        < value["lastWeeklyAt"]
                            .as_i64()
                            .unwrap_or(time.timestamp_millis())
                            .saturating_add(7 * 86400000)
                    || (valid(db, &settings, owner, &value["preview"])?
                        && ["prepared", "running", "ready"]
                            .contains(&string(&value["preview"], "status")))
                {
                    continue;
                }
                write(
                    db,
                    owner,
                    merge(value, &json!({"lastWeeklyAt":time.timestamp_millis()})),
                )?;
                return Ok(prepare_at(db, owner, true, time)
                    .ok()
                    .map(|p| (owner.clone(), string(&p, "id").to_owned())));
            }
            Ok(None)
        })
        .await?;
    if let Some((owner, id)) = work {
        let _ = generate(app, &owner, &id).await;
    }
    Ok(())
}
pub async fn handle(app: &App, ctx: &Context) -> Result<Option<Response>> {
    let route: Vec<_> = ctx.path.iter().map(|s| s.to_ascii_lowercase()).collect();
    let route: Vec<_> = route.iter().map(String::as_str).collect();
    let owner = ctx.owner.clone();
    let body = ctx.body.clone();
    match (ctx.method.as_str(), route.as_slice()) {
        ("POST", ["style", "settings"]) => {
            app.db(move |db| update_settings(db, &owner, &body)).await?
        }
        ("POST", ["style", "preview"]) => {
            app.db(move |db| prepare(db, &owner, false)).await?;
        }
        ("POST", ["style", "generate"]) => {
            generate(app, &owner, string(&body, "previewId")).await?
        }
        ("POST", ["style", "apply"]) => app.db(move |db| apply(db, &owner, &body)).await?,
        ("DELETE", ["style", "profile"]) => app.db(move |db| clear(db, &owner)).await?,
        _ => return Ok(None),
    }
    Ok(Some(
        Json(app.state(&ctx.owner, ctx.paged).await?).into_response(),
    ))
}
